### PR TITLE
WT-5416 Make sure all objects are traversed for rollback to stable

### PR DESCRIPTION
**Pull request by kommiharibabu**
As we are writing the most recent version to the data store, it is
required to scan all the objects for rollback to stable fix up.

We need to read all the internal pages of the objects to fix up.

### CODE REVIEW

File: src/txn/txn_rollback_to_stable.c [hunk 1]
```diff
@@ -432,6 +432,84 @@ __txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
...
+
+    /*
+     * Accumulate errors but continue through to the end of the metadata.
+     */
+    while ((t_ret = cursor->next(cursor)) == 0) {
+        if ((t_ret = cursor->get_key(cursor, &uri)) != 0 || strcmp(uri, WT_METAFILE_URI) == 0 ||
+          strcmp(uri, WT_LAS_URI) == 0) {
+            WT_TRET(t_ret);
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c [hunk 2]
```diff
@@ -468,7 +546,8 @@ __txn_rollback_to_stable(WT_SESSION_IMPL *session, const char *cfg[])
      */
     conn->stable_rollback_maxfile = conn->next_file_id + 1;
     WT_ERR(__bit_alloc(session, conn->stable_rollback_maxfile, &conn->stable_rollback_bitstring));
-    WT_ERR(__wt_conn_btree_apply(session, NULL, __txn_rollback_to_stable_btree, NULL, cfg));
+    WT_WITH_SCHEMA_LOCK(session,
+      ret = __wt_rollback_to_stable_apply_all(session, __txn_rollback_to_stable_btree, NULL, cfg));
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -432,6 +432,84 @@ __txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
...
+        WT_SAVE_DHANDLE(session, WT_TRET(file_func(session, cfg)));
+        WT_TRET(__wt_session_release_dhandle(session));
+    }
+    WT_TRET_NOTFOUND_OK(t_ret);
+
+    return (ret);
+}
+
+/*
+ * __wt_rollback_to_stable_apply_all --
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c [hunk 4]
```diff
@@ -304,7 +304,7 @@ __txn_rollback_to_stable_btree_walk(WT_SESSION_IMPL *session, wt_timestamp_t rol
     /* Walk the tree, marking commits aborted where appropriate. */
     ref = NULL;
     while ((ret = __wt_tree_walk(
-              session, &ref, WT_READ_CACHE | WT_READ_NO_EVICT | WT_READ_WONT_NEED)) == 0 &&
+              session, &ref, WT_READ_CACHE_LEAF | WT_READ_NO_EVICT | WT_READ_WONT_NEED)) == 0 &&
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: I fixed it by changing the ref state to locked and later change it back once the page type is identified.

File: src/txn/txn_rollback_to_stable.c [hunk 5]
```diff
@@ -432,6 +432,84 @@ __txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
...
+
+/*
+ * __wt_rollback_to_stable_apply_all --
+ *     Apply a function to all files listed in the metadata, apart from the metadata and history
+ *     store file.
+ */
+static int
+__wt_rollback_to_stable_apply_all(WT_SESSION_IMPL *session,
+  int (*file_func)(WT_SESSION_IMPL *, const char *[]),
+  int (*name_func)(WT_SESSION_IMPL *, const char *, bool *), const char *cfg[])
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: The existing `__wt_conn_btree_apply()` function that is used by a rollback to stable has the notation of having both `file_func()` and `name_func()`. I followed a similar notation to the new function also.  The `__wt_meta_apply_all()` function also have the similar signature. If it doesn't require, I can change it.

File: src/btree/bt_read.c [hunk 6]
```diff
@@ -247,11 +247,23 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
                 return (WT_NOTFOUND);
             goto read;
         case WT_REF_DISK:
-            /* Limit reads to cache-only, or internal pages only. */
-            if (LF_ISSET(WT_READ_CACHE) ||
-              (LF_ISSET(WT_READ_CACHE_LEAF) && __wt_ref_is_leaf(session, ref)))
+            /* Limit reads to cache-only */
```
Comment (sueloverso) [inline src/btree/bt_read.c]: Minor nit: please end comment with a period at the end of the sentence.

File: src/btree/bt_read.c [hunk 7]
```diff
@@ -247,11 +247,23 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
...
+            if (LF_ISSET(WT_READ_CACHE))
                 return (WT_NOTFOUND);
 
+            /* Limit reads to internal pages only. */
+            if (LF_ISSET(WT_READ_CACHE_LEAF)) {
+                /* Lock the ref to avoid it getting changed, while we check the page type. */
+                if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
+                    return (WT_NOTFOUND);
+
```
Comment (sueloverso) [inline src/btree/bt_read.c]: I suggest removing this blank line because it then keeps all the lock/op/unlock lines together.

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sueloverso) [reply src/txn/txn_rollback_to_stable.c]: I agree with @keithbostic that if no caller uses the arg, then simplify and remove it.

File: src/btree/bt_read.c (hunk 7 above)
Comment (keithbostic) [reply src/btree/bt_read.c]: The fact we were unable to set the state to `WT_REF_LOCKED` doesn't tell us anything about whether or not the page is a leaf page.

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (keithbostic) [reply src/txn/txn_rollback_to_stable.c]: I'll wait on this change to do a second detailed review.

File: src/btree/bt_read.c (hunk 6 above)
Comment (kommiharibabu) [reply src/btree/bt_read.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed by reorganizing the code.

File: src/btree/bt_read.c (hunk 7 above)
Comment (kommiharibabu) [reply src/btree/bt_read.c]: I agree that failure in state change to `WT_REF_LOCKED` led to page read failure. I added a comment explaining the scenario where the `WT_READ_CACHE_LEAF` is used and what happens in case of a failure.

File: src/txn/txn_rollback_to_stable.c [hunk 8]
```diff
@@ -431,9 +429,46 @@ __txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
...
+err:
+    WT_TRET_NOTFOUND_OK(ret);
+    WT_TRET(__wt_metadata_cursor_release(session, &cursor));
+    return (ret);
+}
+
 /*
  * __txn_rollback_to_stable --
- *     Rollback all in-memory state related to timestamps more recent than the passed in timestamp.
+ *     Rollback all related to timestamps more recent than the passed in timestamp.
```
Comment (tammybailey) [inline src/txn/txn_rollback_to_stable.c]: I don't understand this comment. Does it mean "Remove all modifications with timestamps more recent than the passed in timestamp"?

File: src/btree/bt_read.c [hunk 9]
```diff
@@ -247,11 +247,35 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
...
+            /* Limit reads to internal pages only. */
+            if (LF_ISSET(WT_READ_CACHE_LEAF)) {
+                /*
+                 * Currently, the internal page read request is passed only in two scenarios.
+                 *  1. Garbage collection of history store
+                 *  2. Rollback to stable operation
+                 *
+                 * Lock the ref before we check the page type to avoid the ref getting changed
+                 * underneath. Failed to change the ref state to WT_REF_LOCKED can led skip the page
+                 * read.
```
Comment (tammybailey) [inline src/btree/bt_read.c]: This sentence is confusing to me, it feels like there are some words missing: "Failed to change the ref state to WT_REF_LOCKED can led skip the page read." Maybe reword as "Failure to change the ref state to WT_REF_LOCKED can lead to skipping the page read" (if that is the correct interpretation!)

File: src/txn/txn_rollback_to_stable.c (hunk 8 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed with the provided sentence.

File: src/btree/bt_read.c (hunk 9 above)
Comment (kommiharibabu) [reply src/btree/bt_read.c]: Fixed with the provided sentence.

File: src/btree/bt_read.c (hunk 7 above)
Comment (keithbostic) [reply src/btree/bt_read.c]: If we fail to CAS into a locked state, why not restart the loop? The state of the page has changed, why not retry the read? I'm suggesting something like:

```
*** bt_read.c   Thu Jan 23 00:08:42 2020
--- -   Thu Jan 23 00:11:58 2020
*************** 
*** 267,273 ****
                   * a disk page that can led to a failure.
                   */
                  if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
!                     return (WT_NOTFOUND);

                  is_leaf_page = __wt_ref_is_leaf(session, ref);
                  WT_REF_SET_STATE(ref, WT_REF_DISK);
--- 267,273 ----  
                   * a disk page that can led to a failure.
                   */
                  if (!WT_REF_CAS_STATE(session, ref, WT_REF_DISK, WT_REF_LOCKED))
!                     continue;

                  is_leaf_page = __wt_ref_is_leaf(session, ref);
                  WT_REF_SET_STATE(ref, WT_REF_DISK);
```
That way we don't have to care too much about how the function was called.

File: src/txn/txn_rollback_to_stable.c [hunk 10]
```diff
@@ -431,9 +429,46 @@ __txn_rollback_to_stable_check(WT_SESSION_IMPL *session)
...
+        WT_ERR(__wt_session_get_dhandle(session, uri, NULL, NULL, 0));
+        WT_SAVE_DHANDLE(session, ret = __txn_rollback_to_stable_btree(session));
+        WT_TRET(__wt_session_release_dhandle(session));
+        WT_ERR(ret);
+    }
+
+err:
+    WT_TRET_NOTFOUND_OK(ret);
+    WT_TRET(__wt_metadata_cursor_release(session, &cursor));
+    return (ret);
```
Comment (keithbostic) [inline src/txn/txn_rollback_to_stable.c]: I'd move the `WT_TRET_NOTFOUND_OK` check above the `err` label, `WT_NOTFOUND` is only OK at the end of the `cursor.next` loop, it's not necessarily OK after any jump to the `err` label.

File: src/btree/bt_read.c (hunk 7 above)
Comment (kommiharibabu) [reply src/btree/bt_read.c]: A retry is a nice approach, Fixed.

File: src/txn/txn_rollback_to_stable.c (hunk 10 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.
